### PR TITLE
Support drag and drop to arrange the order

### PR DESCRIPTION
**Pull request by tommyme**
🎉 After some refactoring and optimization, the support for this feature has finally been completed 😊

- 🔥 Removed the original functionality of changing order using arrows and **replaced** it with a unified drag-and-drop sorting method.
- 👋 Dragging an item over a node indicates that it will be placed after that node.
- 📌 Drag-and-drop sorting can be toggled using the button in the status bar of the view.

### CODE REVIEW

**Comment by tommyme** [comment]
@Latria-Kure 
What are your thoughts on this change?
If you have time, could you help me with a code review or testing?
Your opinion is very important.

**Comment by Latria-Kure** [comment]
> @Latria-Kure What are your thoughts on this change? If you have time, could you help me with a code review or testing? Your opinion is very important.

When I have free time, I will give you feedback as soon as possible.

**Comment by Latria-Kure** [comment]
I love this feat. It's better than I thought.  Sorting become a global state now. We no longer have to pick and choose which nodes to sort one by one. And sorting by drag and drop is more esaier.
I did some testing and found no issues.  It's wonderful!

File: src/views/bookmark_x/functional_types.ts [hunk 1]
```diff
@@ -498,15 +554,30 @@ export class RootGroup extends Group {
...
+    target_group: Group,
+    strategy: util.AddElStrategy<NodeType> = new util.AddElPushBackStrategy(),
+  ) {
+    if (node instanceof Bookmark) {
+      this.cut_node_recache_all(node);
+      node.uri = target_group.get_full_uri();
+      this.add_node_recache_all(node, strategy);
+    } else if (node instanceof Group) {
+      this.cut_node(node);
+      this.add_node(node, strategy);
```
Comment (tommyme) [inline src/views/bookmark_x/functional_types.ts:568]: before this line, update node.uri, OR, you'll add wrong node!

File: src/views/bookmark_x/bookmark_tree_data_provider.ts [hunk 2]
```diff
@@ -252,3 +140,182 @@ export class BookmarkTreeDataProvider
...
+    this.tvitem = dropping[0];
+    this.item = this.tvitem.base!;
+    this.src_wsf = Controller.get_wsf_with_node(this.item);
+    this.src_rg = Controller.get_root_group(this.src_wsf!);
+    this.dst_rg = this.src_rg;
+    this.target = target;
+    this.flags = new DropFlags(this);
+  }
+  public preprocess() {
+    if (this.flags.Move_To_Undefined) {
```
Comment (tommyme) [inline src/views/bookmark_x/bookmark_tree_data_provider.ts:212]: change this.flags to monitor `node to group`

File: src/views/bookmark_x/bookmark_tree_data_provider.ts [hunk 3]
```diff
@@ -252,3 +140,182 @@ export class BookmarkTreeDataProvider
...
+      // move to it's wsf, monitor a target node for root group
+      this.dst_wsf = this.src_wsf;
+      this.dst_rg = Controller.get_root_group(this.src_wsf);
+      this.target = new BookmarkTreeItem("");
+      this.target.base = Controller.get_root_group(this.dst_wsf);
+    } else if (this.flags.Move_To_WsfTreeItem) {
```
Comment (tommyme) [inline src/views/bookmark_x/bookmark_tree_data_provider.ts:218]: change this.flags to monitor `node to group`

File: src/views/bookmark_x/bookmark_tree_data_provider.ts [hunk 4]
```diff
@@ -252,3 +140,182 @@ export class BookmarkTreeDataProvider
...
+    this.Node_To_Bookmark = handler.target.base instanceof Bookmark;
+    this.Move_To_Group = handler.target.base instanceof Group;
+    this.Move_To_NodeType = this.Move_To_Group || this.Node_To_Bookmark;
+
+    this.Dst_Src_Same = handler.target! && item === handler.target.base!;
+
+    this.Group_To_Group = item instanceof Group && this.Move_To_Group;
+    this.Bookmark_To_Group = item instanceof Bookmark && this.Move_To_Group;
+    this.Is_Sorting = SpaceSortItem.sorting === true;
+    this.Same_Name_Node_In_Target =
```
Comment (tommyme) [inline src/views/bookmark_x/bookmark_tree_data_provider.ts:174]: make sure the flag can't leads to error.

File: src/views/bookmark_x/bookmark_tree_data_provider.ts [hunk 5]
```diff
@@ -252,3 +140,182 @@ export class BookmarkTreeDataProvider
     return bmti;
   }
 }
+
+class DropFlags {
```
Comment (tommyme) [inline src/views/bookmark_x/bookmark_tree_data_provider.ts:144]: consider use arrow function to calc flag when called

File: src/views/bookmark_x/bookmark_tree_data_provider.ts [hunk 6]
```diff
@@ -252,3 +140,182 @@ export class BookmarkTreeDataProvider
...
+  Bookmark_To_NodeType: boolean;
+  Group_To_NodeType: boolean;
+  constructor(handler: DropHandler) {
+    let item = handler.tvitem.base;
+    this.Move_To_Undefined = typeof handler.target === "undefined";
+    this.Move_To_WsfTreeItem = handler.target instanceof WsfTreeItem;
+    // here, if target instanceof WsfTreeItem return false. use tricks to avoid check error.
+    // if you have better tricks, please make a pr ^_^.
+    handler.target = handler.target as BookmarkTreeItem;
+    this.Node_To_Bookmark = handler.target.base instanceof Bookmark;
```
Comment (tommyme) [inline src/views/bookmark_x/bookmark_tree_data_provider.ts:165]: handler.target may be undefined.

**Comment by tommyme** [comment]
things todo before merging:
- [ ] set 'manual' to be default sort method.

**Pull request merged**